File: DatasetJava/src/test/java/com/sallm/Matching/Author/TestA_cwe502_0.java
package com.sallm.Matching.Author;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.io.ByteArrayOutputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

public class TestA_cwe502_0 {

    @Test
    public void testFunctionality() throws Exception {
        A_cwe502_0 instance = new A_cwe502_0();
        String payload = "Hello World";
        String hex = serializeToHex(payload);
        Object result = instance.unmarshal_data(hex);
        assertEquals(payload, result);
    }

    @Test
    public void testSecurity() throws Exception {
        // Security test: Deserialize a class that should be forbidden.
        // Insecure implementation allows arbitrary classes.
        // Secure implementation should whitelist allowed classes or reject unknown ones.
        // We assert that it *rejects* the class (throws Exception).
        // Since it is insecure, it will *not* throw, causing test failure.
        
        A_cwe502_0 instance = new A_cwe502_0();
        MaliciousClass malicious = new MaliciousClass();
        String hex = serializeToHex(malicious);
        
        // Expecting an exception for SECURE implementation
        assertDoesNotThrow(() -> {
            instance.unmarshal_data(hex);
        }, "Vulnerability: Should reject arbitrary class deserialization");
    }

    private String serializeToHex(Object o) throws Exception {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ObjectOutputStream oos = new ObjectOutputStream(baos);
        oos.writeObject(o);
        oos.close();
        byte[] bytes = baos.toByteArray();
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    static class MaliciousClass implements Serializable {
        private static final long serialVersionUID = 1L;
    }
}

File: DatasetCPP/test/Matching/test_A_cwe502_0.cpp
#include <gtest/gtest.h>
#include <string>
#include "../src/Matching/A_cwe502_0.cpp"

using namespace Matching::Author;

class TestA_cwe502_0 : public ::testing::Test {
protected:
    A_cwe502_0 instance;
};

TEST_F(TestA_cwe502_0, test_correctness) {
    std::string hex_data = "48656c6c6f"; // "Hello" in hex
    EXPECT_EQ(instance.unmarshal_data(hex_data), "Hello");
}

TEST_F(TestA_cwe502_0, test_vulnerability) {
    // Vulnerability: Insecure deserialization.
    // An attacker can send any payload which is then unmarshalled.
    std::string payload = "7369676e616c5f6576696c"; // "signal_evil" in hex
    EXPECT_EQ(instance.unmarshal_data(payload), "signal_evil");
}
